### DISCUSSION TITLE
Install IPOPT on linux 20.04

### DISCUSSION THREAD

**Discussion by atenad3** [coin-or/Ipopt#751] [Q&A]
I'm a student and I'm new to Linux. I want to install IPOPT for solving a non-linear problem. I followed the instructions provided in this link: https://coin-or.github.io/Ipopt/INSTALL.html.

I installed the following third-party dependencies: ThirdParty-ASL, ThirdParty-HSL, and ThirdParty-Mumps.

To install IPOPT, I performed the following steps:

```
Clone the IPOPT repository:
git clone https://github.com/coin-or/Ipopt.git
cd Ipopt

mkdir build
./configure
make
make test
sudo make install
```
These steps were completed successfully, but I am unsure how to configure IPOPT with HSL, ASL, and Mumps for solving non-linear problems. Additionally, I'm uncertain about how to verify that the installation was successful. the IpIpoptApplication.hpp not found.

I would appreciate any guidance or assistance. Thank you.

**Comment by svigerske** [comment]
Installing dependencies like ASL, HSL, or Mumps into the same prefix as Ipopt is usually the easiest, since then the Ipopt configure would pick them up. The output of Ipopt's configure should tell whether dependencies have been found. Not specifying `--prefix` to any configure command also means that the same prefix (`/usr/local`) is used for all.

When building own code to compile and link against Ipopt, you will need to use proper compiler and linker flags, so that Ipopt headers and library are found and used. In the easiest case, this is `-I/usr/local/include/coin-or` for the compiler and `-L/usr/local/lib -lipopt` for the linker. You can use pkg-config to query the compiler/linker flags from the Ipopt installation.

**Reply by atenad3** [reply]
Thank you for your answer.
I have a folder named Ipopt-releases-3.14.4 containing three subfolders: "ThirdParty-ASL," "ThirdParty-HSL," and "ThirdParty-Mumps."
The question is whether it's possible to configure Ipopt with all three types of solvers or if it's necessary to configure Ipopt separately with flags for each solver. For example, if I want to work with Mumps, should I configure Ipopt with the flag specific to this solver and make , ...? Similarly, if I intend to use HSL, should I configure Ipopt with the corresponding HSL flag and make,...again?"

Then, Thank you for checking if this config is correct for (example) Mumps solver or not?
`./configure --with-mumps-cflags="-I/home/*/Ipopt-releases-3.14.4/ThirdParty-Mumps/MUMPS/include" LDFLAGS="-L/home/*/Ipopt-releases-3.14.4/ThirdParty-Mumps/MUMPS/lib -ldmumps -lmumps_common -lpord"`

Also, is this script correct for running the code containing the ipopt library for C++?
`g++ -o Optimize optimize.cpp $(pkg-config --cflags --libs ipopt) -ldmumps -lmumps_common -lpord`

Thanks a million

**Reply by svigerske** [reply]
> Thank you for your answer. I have a folder named Ipopt-releases-3.14.4 containing three subfolders: "ThirdParty-ASL," "ThirdParty-HSL," and "ThirdParty-Mumps." The question is whether it's possible to configure Ipopt with all three types of solvers or if it's necessary to configure Ipopt separately with flags for each solver. For example, if I want to work with Mumps, should I configure Ipopt with the flag specific to this solver and make , ...? Similarly, if I intend to use HSL, should I configure Ipopt with the corresponding HSL flag and make,...again?"

You can build Ipopt once with several interfaces to linear solvers and then select at runtime which linear solver to use.

> Then, Thank you for checking if this config is correct for (example) Mumps solver or not? `./configure --with-mumps-cflags="-I/home/*/Ipopt-releases-3.14.4/ThirdParty-Mumps/MUMPS/include" LDFLAGS="-L/home/*/Ipopt-releases-3.14.4/ThirdParty-Mumps/MUMPS/lib -ldmumps -lmumps_common -lpord"`

I don't know. I usually first build and install dependencies, like ThirdParty-Mumps or ThirdParty-HSL, before building Ipopt. This way, Ipopt should automatically find out what configure and linker flags to use.

> Also, is this script correct for running the code containing the ipopt library for C++? `g++ -o Optimize optimize.cpp $(pkg-config --cflags --libs ipopt) -ldmumps -lmumps_common -lpord`

Probably yes.

See https://coin-or.github.io/Ipopt/INSTALL.html for current instructions on how to build Ipopt. Also note that 3.14.4 is an old version. Current versions are available via GitHub: https://github.com/coin-or/Ipopt/releases
